### ISSUE TITLE
Unassigned jobid after checkpoint update

### ISSUE THREAD

**Issue by jmeppley** [snakemake/snakemake#817]
**Snakemake version**
the lastest from github: 5.30.2+6.g674a91ca

**Describe the bug**
After a checkpoint completes in some complicated workflows, an exception is thrown because a jobid is requested for a job that does not have a jobid yet.

**Logs**
This is the exception:
```
Traceback (most recent call last):
  File "/mnt/data0/jmeppley/projects/snakemake_bugs/snakemake.272/DTR-phage-pipeline/snakemake/snakemake/__init__.py", line 687, in snakemake
    success = workflow.execute(
  File "/mnt/data0/jmeppley/projects/snakemake_bugs/snakemake.272/DTR-phage-pipeline/snakemake/snakemake/workflow.py", line 699, in execute
    dag.update_checkpoint_dependencies()
  File "/mnt/data0/jmeppley/projects/snakemake_bugs/snakemake.272/DTR-phage-pipeline/snakemake/snakemake/dag.py", line 1300, in update_checkpoint_dependencies
    logger.info("Updating job {} ({}).".format(self.jobid(j), j))
  File "/mnt/data0/jmeppley/projects/snakemake_bugs/snakemake.272/DTR-phage-pipeline/snakemake/snakemake/dag.py", line 704, in jobid
    return self._jobid[job]
KeyError: align_dtr_seqs_to_polished_refs
```

**Minimal example**
A reproducible way to generate this error is to download the [DTR-phage-pipeline](https://github.com/nanoporetech/DTR-phage-pipeline) and run the test. Unfortunately, I have not been able to reproduce the error with a simpler workflow.

**Comment by pb-cdunn** [comment]
```
Snakemake version
the lastest from github: 5.30.2+6.g674a91ca
```
The fix was #727 was not in `5.30.2`, but it's in `5.31.1`. Have you tried the latest version from PyPI? Or the tip of `master` here?

I think a test was added for that other fix (mentioned in your #818). So if you revert that fix and apply your own, you might have a better solution.

**Comment by jmeppley** [comment]
@pb-cdunn Yes, this was found in the development branch for the #727 bugfix (#801). I reported the version wrong, I think.

I actually did try reverting 71b5aeb and applying my fix, but it seems to slow things down. Not surprisingly (in retrospect) assigning all the jobids at once is significantly faster, but somehow a couple jobs slip through in rare cases, so adding a fall back fixed this bug for me.

It may be better to figure out why a few jobs are slipping through the cracks, but I've had no luck yet.
